src/operators/join.rs

```diff
@@ -320,100 +320,127 @@ impl<G, T1> JoinCore<G, T1::Key, T1::Val, T1::R> for Arranged<G,T1>
-
-        let mut input1_buffer = Vec::new();
-        let mut input2_buffer = Vec::new();
-
-        self.stream.binary_frontier(&other.stream, Pipeline, Pipeline, "Join", move |_cap, info| {
+        self.stream.binary_frontier(&other.stream, Pipeline, Pipeline, "Join", move |capability, info| {
 
+            // Acquire an activator to reschedule the operator when it has unfinished work.
             use timely::scheduling::Activator;
             let activations = self.stream.scope().activations().clone();
             let activator = Activator::new(&info.address[..], activations);
 
+            // Our initial invariants are that for each trace, physical compaction is less or equal the trace's upper bound.
+            // These invariants ensure that we can reference observed batch frontiers from `_start_upper` onward, as long as
+            // we maintain our physical compaction capabilities appropriately.
+
+            // I can't say with confidence what should happen if this isn't the case, as it means we are unable to subset
+            // the current data without waiting for data that have not yet been created.
+            let mut trace1_start_upper = Antichain::new();
+            trace1.read_upper(&mut trace1_start_upper);
+            assert!(PartialOrder::less_equal(&trace1.get_physical_compaction(), &trace1_start_upper.borrow()));
+
+            let mut trace2_start_upper = Antichain::new();
+            trace2.read_upper(&mut trace2_start_upper);
+            assert!(PartialOrder::less_equal(&trace2.get_physical_compaction(), &trace2_start_upper.borrow()));
+
+            // acknowledged frontier for each input.
+            use timely::progress::frontier::Antichain;
+            let mut acknowledged1 = Antichain::from_elem(<G::Timestamp>::minimum());
+            let mut acknowledged2 = Antichain::from_elem(<G::Timestamp>::minimum());
+
+            // deferred work of batches from each input.
+            let mut todo1 = std::collections::VecDeque::new();
+            let mut todo2 = std::collections::VecDeque::new();
+
+            // We'll unload the initial batches here, to put ourselves in a less non-deterministic state to start.
+            trace1.map_batches(|batch1| {
+                acknowledged1.clone_from(batch1.upper());
+                // No `todo1` work here, because we haven't accepted anything into `batches2` yet.
+                // It is effectively "empty", because we choose to drain `trace1` before `trace2`.
+                // Once we start streaming batches in, we will need to respond to new batches from
+                // `input1` with logic that would have otherwise been here. Check out the next loop
+                // for the structure.
+            });
+
+            trace2.map_batches(|batch2| {
+                acknowledged2.clone_from(batch2.upper());
+                // It is safe to ask for `ack1` because we have confirmed it to be in advance of `distinguish_since`.
```

Comment:
It's not clear to me where we confirmed that

Edit: it seems like `acknowledged1` will be the upper bound of the "most recent" batch in `trace1` and maybe that is how we know its in advance of `distinguish_since`?

Reply:
Good point. We should be able to assert just after the loading, that the trace upper bounds are exactly the acknowledged frontiers. They are read out the same way (by scanning the existing batches), but that's not clear in this code.